src/core/tsi/alts_transport_security.cc

```diff
@@ -50,6 +50,7 @@ void grpc_tsi_alts_init() {
   g_alts_resource.is_cq_drained = false;
   gpr_mu_init(&g_alts_resource.mu);
   gpr_cv_init(&g_alts_resource.cv);
+  gpr_ref_init(&g_alts_resource.refcount, 0);
 }
 
 void grpc_tsi_alts_shutdown() {
```

Comment:
If this gets called, then we will shut down the resources even if the refcount hasn't gone to zero.

Reply:
Will this ever happen? Credentials will be destroyed before grpc_shutdown is called.

If it is possible grpc_shutdown is called before refcount > 0, we still did not solve root cause of crash.

Reply:
If that's the case, then shouldn't this code simply assert that `g_alts_resource.cq == nullptr`?

My point here is that we shouldn't have two paths to make this shutdown happen.  If the refcount should always be zero by the time we get here, then there's no reason to ever try to do the shutdown here.

Reply:
Another reason for the existence of this chunk of code is that Envoy does will not use ref-count (because they do not use any credential object), so it still relies on `grpc_tsi_alts_shutdown` to destroy channels, cq and so on.

Reply:
I don't think it makes sense to have an API that can be used in both a ref-counted and non-ref-counted fashion.  We should pick one of the two approaches and stick with it.  If this is going to be ref-counted and Envoy is going to use it, then Envoy should be responsible for taking a ref and releasing it when it is no longer needed.

Reply:
It seems non-trivial to let Envoy to use ref-count since they directly use `alts_tsi_handshaker`, and it does not make sense to ref (or unref) shared resources in TSI handshakers as they need to be associated with credential objects. Do you have any suggestion Mark? If not, I will discard this PR and directly work on removing TSI threads.

Reply:
Why is it hard for Envoy to do this?  They would just need to take a single ref at startup and unref it at shutdown.

Even if we remove the TSI thread for the gRPC use-case, Envoy will still need its own thread, so I don't think that will solve the problem.